polygon/heimdall/span_block_producers_tracker.go

```diff
@@ -0,0 +1,172 @@
+package heimdall
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"sync/atomic"
+
+	"github.com/erigontech/erigon-lib/log/v3"
+	"github.com/erigontech/erigon/polygon/bor/borcfg"
+	"github.com/erigontech/erigon/polygon/bor/valset"
+)
+
+func newSpanBlockProducersTracker(
+	logger log.Logger,
+	borConfig *borcfg.BorConfig,
+	store EntityStore[*SpanBlockProducerSelection],
+) *spanBlockProducersTracker {
+	return &spanBlockProducersTracker{
+		logger:     logger,
+		borConfig:  borConfig,
+		store:      store,
+		newSpans:   make(chan *Span),
+		idleSignal: make(chan struct{}),
+	}
+}
+
+type spanBlockProducersTracker struct {
+	logger     log.Logger
+	borConfig  *borcfg.BorConfig
+	store      EntityStore[*SpanBlockProducerSelection]
+	newSpans   chan *Span
+	queued     atomic.Int32
+	idleSignal chan struct{}
+}
+
+func (t *spanBlockProducersTracker) Run(ctx context.Context) error {
+	defer close(t.idleSignal)
+
+	for {
+		select {
+		case <-ctx.Done():
+			return ctx.Err()
+		case newSpan := <-t.newSpans:
+			err := t.ObserveSpan(ctx, newSpan)
+			if err != nil {
+				return err
+			}
+
+			t.queued.Add(-1)
```

Comment:
What is the -1 for?

Reply:
to indicate that 1 item from the newSpan queue has been processed